[WIP] Implement Time Signatures section of Inspector

shoogle opened the pull request:
![image](https://user-images.githubusercontent.com/11011881/81266166-303d9080-903c-11ea-9b9a-083226960f3e.png)

The "Change time signature" and "Properties" buttons don't do anything at the moment.

shoogle committed:
Implement model for Time Signatures section of Inspector

Harmoniker1 commented on inspectors/models/notation/timesignature/timesignaturesettingsmodel.cpp:
```diff
@@ -0,0 +1,53 @@
+
+void TimeSignatureSettingsModel::createProperties()
+{
+    m_hasToShowCourtesy = buildPropertyItem(Ms::Pid::SHOW_COURTESY);
```
`hasToShowCourtesy`? Why not just `showCourtesy`?

shoogle replied:
Consistency with [keysignaturesettingsmodel.cpp](https://github.com/vpereverzev/MuseScore/blob/0dd6aa0ba73f93802548802c6cdbe5d2efea171c/inspectors/models/notation/keysignature/keysignaturesettingsmodel.cpp#L13), but I agree that `showCourtesy` is better assuming this a a simple `on`/`off` switch rather than `auto`/`force` which "hasTo" would imply.

vpereverzev replied:
"is" and "has" are usually markers that it's a bool property
In QML, "showCourtesy" may sound like a direct action - i.e. "Q_INVOKABLE void showCourtesy()"

Harmoniker1 commented on inspectors/types/timesignaturetypes.h:
```diff
@@ -0,0 +1,27 @@
+{
+    Q_GADGET
+
+    Q_ENUMS(Mode)
```
Please use `Q_ENUM` instead of `Q_ENUMS` as the latter is deprecated. Also, please put `Q_ENUM` after the enum type declaration.

shoogle replied:
Thanks, but I don't think this file is actually needed at all. I was using the key signatures files as a template for the time signature implementation, but they are not exactly the same.

Howard-C commented on inspectors/types/timesignaturetypes.h:
```diff
@@ -0,0 +1,27 @@
+{
+    Q_GADGET
+
+    Q_ENUMS(Mode)
```
Please use `Q_ENUM` instead of `Q_ENUMS` as the latter is deprecated. Also, please put `Q_ENUM` after the enum type declaration.

Harmoniker1 commented on inspectors/types/timesignaturetypes.h:
```diff
@@ -0,0 +1,27 @@
+    Q_ENUMS(Mode)
+
+public:
+    enum Mode {
```
Please use `enum class`.

Harmoniker1 commented on inspectors/view/qml/notation/timesignatures/TimeSignaturePopup.qml:
```diff
@@ -0,0 +1,101 @@
+import QtQuick.Layouts 1.3
+import "../../common"
+
+StyledPopup {
+    id: root
+
+    property QtObject model: null
+
+    implicitHeight: contentColumn.implicitHeight + topPadding + bottomPadding
+    width: parent.width
+
```
There should be an icon property around here. Is the icon not ready yet?

shoogle replied:
I don't see one in [KeySignaturePopup.qml](https://github.com/vpereverzev/MuseScore/blob/0dd6aa0ba73f93802548802c6cdbe5d2efea171c/inspectors/view/qml/notation/keysignatures/KeySignaturePopup.qml).

vpereverzev replied:
I've simplified the structure of proxy models and the new popups declaration at the beginning of the week

Harmoniker1 reviewed (commented):
The comments on `timesignaturetypes.h` are based on [Qt documentation](https://doc.qt.io/qt-5/qobject.html#Q_ENUM) and the same problems for other enum type declarations are fixed in #10. These are the things Vasily and I agreed upon.

shoogle commented:
Status update: I have fixed bugs and implemented the Properties button so it triggers the existing Time Signature Properties dialog.

![timesig-properties-dialog](https://user-images.githubusercontent.com/11011881/81586862-1db1b700-93ae-11ea-8a03-9999a7e7a1ac.jpg)

However, there is no existing dialog for "Change time signature". I could create one, but I have run into a problem with scope. This issue is whether changes in time signature or its properties should apply to all staves or only the current one? (Related: should selecting a time signature in one staff also select those in other staves too? See [issue #29826](https://musescore.org/en/node/29826#comment-155866) for historic context.) These require tweaks to the design. I'm in discussion with @Tantacrul.

Tantacrul commented:
Hey Peter,

Sorry I didn’t get around to your questions today. Let me write up a quick
idea of what I think it should do and I’ll send it over shortly.

I guess it’s easy for me to point to what other apps do but I suspect from
your question, there a additional complexity to be considered.

So... what I would definitely imagine should happen is that when a user
clicks on a time signature, say 3/4, And they choose to change it to 2/4,
That means that all instruments on that bar have now also been changed to
2/4 too. But any other instances of time changes either before or after
that instance would not be affected. Now, I’m not sure if this is even
possible, but If there is an instrument that is using a contrary time
signature, then it should remains as it is.

I’ll drop a quick description of what I imagine should happen later on
tonight and I’ll send it tomorrow morning.

Thanks a lot,
Martin

On Mon, 11 May 2020 at 17:49, Peter Jonas <notifications@GitHub.com> wrote:

> Status update: I have fixed bugs and implemented the Properties button so
> it triggers the existing Time Signature Properties dialog.
>
> [image: timesig-properties-dialog]
> <https://user-images.githubusercontent.com/11011881/81586862-1db1b700-93ae-11ea-8a03-9999a7e7a1ac.jpg>
>
> However, there is no existing dialog for "Change time signature". I could
> create one, but I have run into a problem with scope. This issue is whether
> changes in time signature or its properties should apply to all staves or
> only the current one? (Related: should selecting a time signature in one
> staff also select those in other staves too? See issue #29826
> <https://musescore.org/en/node/29826#comment-155866> for historic
> context.) These require tweaks to the design. I'm in discussion with
> @Tantacrul <https://github.com/Tantacrul>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vpereverzev/MuseScore/pull/11#issuecomment-626821146>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFAMMZZQQLC4OBXDEWL6WADRRAUC3ANCNFSM4M3CNRXA>
> .
>

shoogle committed:
Implement view for Time Signatures section of Inspector

shoogle commented:
Changed method of triggering Properties dialog to match 44ab4963a5. View has been changed to match the new design by @Tantacrul which combines "Properties" and "Change time signature" with a single button.

![timesig_inspector](https://user-images.githubusercontent.com/11011881/83030900-d53ffd80-a02b-11ea-934d-8bdeb6fd0af6.png)

However, further changes are needed to make the dialog affect more than one staff.

Harmoniker1 commented:
I think this one can be closed?
